collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java

```diff
@@ -69,32 +73,57 @@ public class FormHierarchyFragment extends Fragment {
      * was accessed. Used to jump the user back to where they were if applicable.
      */
     private FormIndex startIndex;
-
-    public FormHierarchyFragment(boolean viewOnly, ViewModelProvider.Factory viewModelFactory, MenuHost menuHost) {
+    private final Scheduler scheduler;
+    private final InstancesDataService instancesDataService;
+    private final String currentProjectId;
+
+    public FormHierarchyFragment(
+            boolean viewOnly,
+            ViewModelProvider.Factory viewModelFactory,
+            MenuHost menuHost,
+            Scheduler scheduler,
+            InstancesDataService instancesDataService,
+            String currentProjectId
+    ) {
         super(R.layout.form_hierarchy_layout);
         this.viewOnly = viewOnly;
-
         this.viewModelFactory = viewModelFactory;
         this.menuHost = menuHost;
+        this.scheduler = scheduler;
+        this.instancesDataService = instancesDataService;
+        this.currentProjectId = currentProjectId;
     }
 
     @Override
     public void onAttach(@NonNull Context context) {
         super.onAttach(context);
 
         formEntryViewModel = new ViewModelProvider(requireActivity(), viewModelFactory).get(FormEntryViewModel.class);
-        formHierarchyViewModel = new ViewModelProvider(this, new ViewModelProvider.Factory() {
-            @NonNull
-            @Override
-            public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
-                return (T) new FormHierarchyViewModel();
-            }
-        }).get(FormHierarchyViewModel.class);
+        formHierarchyViewModel = new ViewModelProvider(
+                this,
+                new FormHierarchyViewModel.Factory(scheduler, currentProjectId, instancesDataService)
+        ).get(FormHierarchyViewModel.class);
         requireActivity().setTitle(formEntryViewModel.getFormController().getFormTitle());
-
         startIndex = formEntryViewModel.getFormController().getFormIndex();
 
+        MaterialProgressDialogFragment.showOn(this, formHierarchyViewModel.isCloning(), getParentFragmentManager(), () -> {
+            MaterialProgressDialogFragment dialog = new MaterialProgressDialogFragment();
+            dialog.setMessage(getString(org.odk.collect.strings.R.string.preparing_form_edit));
+            return dialog;
+        });
+
         menuProvider = new FormHiearchyMenuProvider(formEntryViewModel, formHierarchyViewModel, viewOnly, new FormHiearchyMenuProvider.OnClickListener() {
+            @Override
+            public void onEditClicked() {
+                formHierarchyViewModel
```

Comment:
Could this code just live in the menu provider and avoid needing the callback?

Reply:
We would still need some callbacks. It's not just that `onEditClicked` event to handle but also `onGoUpClicked`, `onAddRepeatClicked`, `onDeleteRepeatClicked`. Some of them for example need to update the hierarchy etc.